File: LeetCode/python3/127.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        n, count = len(beginWord), 0
        dict = {}
        for word in wordList:
            dict[word] = 1
        queue = [beginWord]
        size = 1
        word = beginWord
        while len(queue) > 0:
            word = queue[0]
            if word == endWord:
                break
            queue = queue[1:]
            for i in range(len(word)):
                temp = word
                for j in range(26):
                    if chr(ord('a') + j) == word[i]:
                        continue
                    temp = temp[:i] + chr(ord('a') + j) + temp[i + 1:]
                    if temp in dict:
                        queue.append(temp)
                        del dict[temp]
            size -= 1
            if size == 0:
                size = len(queue)
                count += 1
        return count + 1 if word == endWord else 0

File: LeetCode/cpp/127.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string> &wordList) {
        int n = beginWord.size(), count = 0;
        if (n != endWord.size())
            return 0;
        unordered_map<string, bool> words;
        for (auto &l:wordList)
            words[l] = true;
        queue<string> que;
        int size = 1;
        que.push(beginWord);
        string word = que.front();
        while (!que.empty()) {
            word = que.front();
            if (word == endWord)
                break;
            que.pop();
            for (int i = 0; i < word.size(); ++i) {
                string temp = word;
                for (int j = 0; j < 26; ++j) {
                    if ('a' + j == word[i])
                        continue;
                    temp[i] = 'a' + j;
                    if (words.find(temp) != words.end()) {
                        que.push(temp);
                        words.erase(temp);
                    }
                }
            }
            --size;
            if (size == 0) {
                size = que.size();
                ++count;
            }
        }
        return word == endWord ? count + 1 : 0;
    }
};